lib/widgets/compose_box.dart

```diff
@@ -28,10 +28,30 @@ const double _composeButtonSize = 44;
 ///
 /// Subclasses must ensure that [_update] is called in all exposed constructors.
 abstract class ComposeController<ErrorT> extends TextEditingController {
+  int get maxLengthUnicodeCodePoints;
+
   String get textNormalized => _textNormalized;
   late String _textNormalized;
   String _computeTextNormalized();
 
+  /// Length of [textNormalized] in Unicode code points,
+  /// if it might exceed [maxLengthUnicodeCodePoints], else null.
+  ///
+  /// Use this instead of [String.length]
+  /// to enforce a max length expressed in code points.
+  /// [String.length] is conservative and may cut the user off too short.
+  ///
+  /// Counting code points ([String.runes])
+  /// is more expensive than getting the number of UTF-16 code units
+  /// ([String.length]), so we avoid it when the result definitely won't exceed
+  /// [maxLengthUnicodeCodePoints].
+  int? get lengthUnicodeCodePointsIfLong => _lengthUnicodeCodePointsIfLong;
```

Comment:
What if we maintain the length internally, and expose only a method that tests if the content is too long? We already have access to the length limit via `maxLengthUnicodeCodePoints` and hiding this cache might simplify the checks a bit.

Reply:
Sure, makes sense; I can remove this public getter.

> and expose only a method that tests if the content is too long

How about keeping the `validationErrors` getter as the way consumers get this information, without adding a new method?